File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-3.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        l = res = max_f = 0
        map = {}

        for r in range(len(s)):
            map[s[r]] = 1 + map.get(s[r], 0)
            max_f = max(max_f, map[s[r]])

            while (r - l - max_f) + 1 > k:
                map[s[l]] -= 1
                l += 1
            
            res = max(res, r - l + 1)
        
        return res

File: Data Structures & Algorithms/binary-tree-diameter/submission-3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int res;
    public int diameterOfBinaryTree(TreeNode root) {
        maxDepth(root);
        return res;
    }

    private int maxDepth(TreeNode node){
        if(node == null) return 0;
        int left = maxDepth(node.left);
        int right = maxDepth(node.right);
        res = Math.max(res, left + right);
        return 1 + Math.max(left, right);
    }
}
